lib/cli/watch-run.js

```diff
@@ -0,0 +1,106 @@
+'use strict';
+
+const utils = require('../utils');
+const Context = require('../context');
+const Mocha = require('../mocha');
+
+/**
+ * Exports the `watchRun` function that runs mocha in "watch" mode.
+ * @see module:lib/cli/run-helpers
+ * @module
+ * @private
+ */
+
+/**
+ * Run Mocha in "watch" mode
+ * @param {Mocha} mocha - Mocha instance
+ * @param {Object} opts - Options
+ * @param {string[]} [opts.extension] - List of extensions to watch
+ * @param {string|RegExp} [opts.grep] - Grep for test titles
+ * @param {string} [opts.ui] - User interface
+ * @param {string[]} [opts.files] - Array of test files
+ * @param {string} [opts.cwd] - Current working directory
+ * @private
+ */
+module.exports = (mocha, {extension, grep, ui, files, cwd}) => {
+  let runner;
+
+  console.log();
+  hideCursor();
+  process.on('SIGINT', () => {
+    showCursor();
+    console.log('\n');
+    process.exit(130);
+  });
+
+  const watchFiles = utils.files(cwd, extension);
+  let runAgain = false;
+
+  const loadAndRun = () => {
+    try {
+      mocha.files = files;
+      runAgain = false;
+      runner = mocha.run(() => {
+        runner = null;
+        if (runAgain) {
+          rerun();
+        }
+      });
+    } catch (e) {
+      console.log(e.stack);
+    }
+  };
+
+  const purge = () => {
+    watchFiles.forEach(Mocha.unloadFile);
+  };
+
+  loadAndRun();
+
+  const rerun = () => {
+    purge();
```

Comment:
Would think there should be a `clearScreen` here... see #2312

Reply:
I’d like to avoid changing behavior in this PR. It is also not clear to me which behavior we exactly want.